SIPS/sip-21.md

```diff
@@ -0,0 +1,53 @@
+---
+sip: 21
+title: Snap-defined timeouts
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/128
+author: Frederik Bolding (@frederikbolding)
+created: 2024-01-12
+---
+
+## Abstract
+
+This SIP describes a way for Snaps to customize the request timeouts that constrain the Snap lifecycle. Effectively allowing the Snap to opt-in to slightly longer or shorter timeouts where needed.
+
+## Motivation
+
+Snaps that have to do CPU intensive computation are currently limited by the default `1` minute timeout, causing certain use-cases to be unsupportable by Snaps (e.g. computing ZK proofs). By letting Snaps customize the request timeout (within reason) for each type of handler that they expose, the Snap developer gets more control over the user experience while the Snap platform lifecycle requirements can remain strict.
+
+## Specification
+
+> Formal specifications are written in TypeScript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP specifies an addendum that MUST be applied to existing and future Snap handler permissions (`endowment:rpc`, `endowment:transaction-insight` etc.). The addendum is an optional caveat value called `maxRequestTime`.
+
+The value MUST be a valid integer between `5000` (5 seconds) and `180000` (3 minutes) and specifies the request timeout for the given handler in milliseconds. If no value is provided the default timeout of 1 minute (`30000` ms) MUST be used.
```

Comment:
Is 3 minutes enough for the ZK Snaps?

Reply:
I think so, but we will get some feedback from snap devs before proceeding!

Reply:
Yes my understanding is 3 minutes would be enough.

Reply:
From Rarimo: "3 minutes should be enough but we may want to do up to 5 minutes for more complex ZKPs in the future."

"Usually it takes about 40s on average hardware and up to 2 minutes on slow devices to generate a ZKP at https://robotornot.rarimo.com/" 

"We’ll have estimates for a bit more complex ZKPs in our new project in the next two weeks"